ResearchKit/ActiveTasks/ORKTowerOfHanoiStepViewController.m

```diff
@@ -0,0 +1,275 @@
+    [super viewWillDisappear:animated];
+    [_timer invalidate];
+}
+
+- (void)updateViewConstraints {
+    [super updateViewConstraints];
+    if (_currentConstraints) {
+        [NSLayoutConstraint deactivateConstraints:_currentConstraints];
+        _currentConstraints = nil;
+    }
+    BOOL needCompactLayout = self.traitCollection.horizontalSizeClass == UIUserInterfaceSizeClassCompact &&
+    self.traitCollection.verticalSizeClass != UIUserInterfaceSizeClassCompact;
+    _currentConstraints = needCompactLayout ? [self compactConstraints] : [self regularConstraints];
+    [NSLayoutConstraint activateConstraints:_currentConstraints];
+}
+
+#pragma Mark -- ORKStepViewController
+
+- (void)skipForward {
+    [self finish];
+}
+
+#pragma Mark -- ORKActiveTaskViewController
+
+- (ORKResult *)result {
+    ORKTowerOfHanoiResult *result = [[ORKTowerOfHanoiResult alloc] initWithIdentifier:self.step.identifier];
+    result.moves = self.moves;
+    result.puzzleWasSolved = [self puzzleIsSolved];
+    if (_firstMoveDate != nil) {
+        result.startDate = _firstMoveDate;
+    }
+    return result;
+}
+
+#pragma Mark -- ORKTowerOfHanoiTowerViewDataSource
+ 
+- (NSUInteger)numberOfDisksInTowerOfHanoiView:(ORKTowerOfHanoiTowerView *)towerView {
+     NSInteger towerIndex = [_towerViews indexOfObject:towerView];
+     ORKTowerOfHanoiTower *tower = _towers[towerIndex];
+     return tower.disks.count;
+}
+ 
+- (NSNumber *)towerOfHanoiView:(ORKTowerOfHanoiTowerView *)towerView diskAtIndex:(NSUInteger)index {
+     NSInteger towerIndex = [_towerViews indexOfObject:towerView];
+     ORKTowerOfHanoiTower *tower = _towers[towerIndex];
+     return tower.disks[index];
+}
+
+#pragma Mark -- ORKTowerOfHanoiTowerViewDelegate
+
+- (void)towerOfHanoiTowerViewWasSelected:(ORKTowerOfHanoiTowerView *)towerView {
+    NSInteger newSelectedIndex = [_towerViews indexOfObject:towerView];
+    if (_selectedIndex == nil) {
+        _selectedIndex = @(newSelectedIndex);
+    }
+    else if ([_selectedIndex isEqual:@(newSelectedIndex)]) {
+        _selectedIndex = nil;
+    } else {
+        [self transferDiskFromTowerAtIndex:_selectedIndex.integerValue toTowerAtIndex:newSelectedIndex];
+    }
+    [self reloadData];
+    [self evaluatePuzzle];
+}
+
+#pragma Mark -- ORKTowerOfHanoiViewController
+
+- (NSMutableArray *)moves {
+    if (_moves) {
+        return _moves;
+    }
+    _moves = [NSMutableArray array];
+    return _moves;
+}
+
+- (NSDateComponentsFormatter *)dateComponentsFormatter {
+    if (_dateComponentsFormatter) {
+        return _dateComponentsFormatter;
+    }
+    _dateComponentsFormatter = [NSDateComponentsFormatter new];
+    _dateComponentsFormatter.unitsStyle = NSDateComponentsFormatterUnitsStylePositional;
+    _dateComponentsFormatter.zeroFormattingBehavior = NSDateComponentsFormatterZeroFormattingBehaviorPad;
+    _dateComponentsFormatter.allowedUnits =  NSCalendarUnitMinute | NSCalendarUnitSecond;
+    return _dateComponentsFormatter;
```

Comment:
This could save a `return` :)
```
if (_dateComponentsFormatter == nil) {
_dateComponentsFormatter = [NSDateComponentsFormatter new];
_dateComponentsFormatter.unitsStyle = NSDateComponentsFormatterUnitsStylePositional;
_dateComponentsFormatter.zeroFormattingBehavior = NSDateComponentsFormatterZeroFormattingBehaviorPad;
_dateComponentsFormatter.allowedUnits =  NSCalendarUnitMinute | NSCalendarUnitSecond;
}
return _dateComponentsFormatter;
```